swiftsimio/subset_writer.py

```diff
@@ -0,0 +1,172 @@
+        the name of the dataset we're interested in. This is the name from the
+        hdf5 file (i.e. "PartType0", rather than "gas")
+
+    Returns
+    -------
+    np.ndarray
+        mask for the appropriate dataset
+
+    """
+    legend = {"Cells": 0, 
+              "PartType0": mask.gas if hasattr(mask, "gas") else 0, 
+              "PartType1": mask.dark_matter if hasattr(mask, "dark_matter") else 0, 
+              "PartType2": 0, 
+              "PartType3": 0, 
+              "PartType4": mask.stars if hasattr(mask, "stars") else 0, 
+              "PartType5": mask.black_holes if hasattr(mask, "black_holes") else 0}
+    for key in legend.keys():
+        if key in dataset_name:
+            return legend[key]
+
+def write_datasubset(infile, outfile, mask: SWIFTMask, dataset_names):
+    """
+    Writes subset of all datasets contained in snapshot according to specified mask
+    Parameters
+    ----------
+    infile : h5py.File
+        hdf5 file handle for input snapshot
+    outfile : h5py.File
+        hdf5 file handle for output snapshot
+    mask : SWIFTMask
+        the mask used to define subset that is written to new snapshot
+    dataset_names : list of str
+        names of datasets found in the snapshot
+    """
+    if mask is not None:
+        for name in dataset_names:
+            if "Cells" not in name:
+                # get output dtype and size 
+                first_value = infile[name][0]
+                output_type = first_value.dtype
+                output_size = first_value.size
+                mask_size = compute_mask_size(mask, name)
+                if output_size != 1:
+                    output_shape = (mask_size, output_size)
+                else:
+                    output_shape = mask_size
+    
+                dataset_mask = get_mask_label(mask, name)
+                subset = read_ranges_from_file(infile[name], dataset_mask, output_shape = output_shape, output_type = output_type)
+                
+                # Write the subset
+                outfile.create_dataset(name, data=subset)
+
+def write_metadata(infile, outfile):
+    """
+    Copy over all the metadata from snapshot to output file
+
+    Parameters
+    ----------
+    infile : h5py.File
+        hdf5 file handle for input snapshot
+    outfile : h5py.File
+        hdf5 file handle for output snapshot
+    """
+    for field in infile.keys():
+        if not any([group_str in field for group_str in ["PartType"]]):
+            infile.copy(field, outfile)
+
+def find_datasets(input_file: h5py.File, dataset_names, path = None):
+    """
+    Recursively finds all the datasets in the snapshot and writes them to a list
+
+    Parameters
+    ----------
+    input_file : h5py.File
+        hdf5 file handle for snapshot
+    dataset_names : list of str
+        names of datasets found in the snapshot
+    path : str, optional
+        the path to the current location in the snapshot
+    """
+    if path != None:
```

Comment:
Never check for equality with `None` like this, always use `is not None` - see https://stackoverflow.com/questions/2209755/python-operation-vs-is-not